client/question/services/FeedbackGeneratorService.js

```diff
@@ -263,34 +328,75 @@ tie.factory('FeedbackGeneratorService', [
      * @param {string|null} outputFunctionName If needed, is string of the
      *    name for the function needed to process/format the user's output.
      * @param {CorrectnessTest} TestCase
+     * @param {number} testCaseIndex
      * @param {*} observedOutput Actual output for running user's code.
+     * @param {string} testSuiteId
      * @returns {Feedback}
      * @private
      */
     var _getCorrectnessTestFeedback = function(
-      outputFunctionName, testCase, observedOutput) {
+        outputFunctionName, testCase, testCaseIndex, observedOutput,
+        testSuiteId) {
+      var testCaseKey = testSuiteId + '-' + testCaseIndex;
       var allowedOutputExample = testCase.getAnyAllowedOutput();
       var feedback = FeedbackObjectFactory.create(
         FEEDBACK_CATEGORIES.INCORRECT_OUTPUT_FAILURE);
-      // TODO(eyurko): Add varied statements for when code is incorrect.
-      feedback.appendTextParagraph('Your code produced the following result:');
-      if (outputFunctionName) {
-        feedback.appendCodeParagraph(
-          'Input: ' + _jsToHumanReadable(testCase.getInput()) + '\n' +
-          ('Output (after running ' + outputFunctionName + '): ' +
-            _jsToHumanReadable(observedOutput)));
-        feedback.appendTextParagraph(
-          'However, the expected output of ' + outputFunctionName + ' is:');
-      } else {
-        feedback.appendCodeParagraph(
-          'Input: ' + _jsToHumanReadable(testCase.getInput()) + '\n' +
-          'Output: ' + _jsToHumanReadable(observedOutput));
-        feedback.appendTextParagraph('However, the expected output is:');
+      // Check if new / next test suite
+      if (testSuiteId != previousTestSuiteId) {
+        previousTestSuiteId = testSuiteId;
+        // Catch regressions
+        if (correctnessTestStates.hasOwnProperty(testCaseKey)) {
+          feedback.appendTextParagraph(
+            'It looks like there was a regression in your code. Your code ' +
+            'worked for the following, but it now fails:');
+          feedback.appendCodeParagraph(
+            'Input: ' + _jsToHumanReadable(testCase.getInput()) + '\n' +
+            'Expected Output: ' + _jsToHumanReadable(allowedOutputExample));
+          return feedback;
+        }
+        correctnessTestStates[testCaseKey] = CORRECTNESS_STATE_STARTING;
+      }
+      // Check if sample input test suite (input and expected already displayed)
+      if (testSuiteId == TEST_SUITE_ID_SAMPLE_INPUT) {
+        correctnessTestStates[testCaseKey] =
+          CORRECTNESS_STATE_EXPECTED_OUTPUT_DISPLAYED;
+      }
+      switch(correctnessTestStates[testCaseKey]) {
+        case CORRECTNESS_STATE_STARTING:
+          // Display input user should try
+          feedback.appendTextParagraph(
+            _getCorrectnessFeedbackString('INPUT_TO_TRY'));
```

Comment:
It would be nice if the string literals here could be replaced with the already-named CORRECTNESS_STATE_* constants (and the object constant modified accordingly); we should also add a Karma test to ensure that the keys of CORRECTNESS_FEEDBACK_TEXT match the existing constants.